File: ios/Classes/ScannerPlatformView.swift
import AVFoundation
import Flutter
import UIKit

/// A `UIView` that keeps the `AVCaptureVideoPreviewLayer` sized to its bounds.
final class PreviewContainerView: UIView {
  private let previewLayer: AVCaptureVideoPreviewLayer
  init(previewLayer: AVCaptureVideoPreviewLayer) {
    self.previewLayer = previewLayer
    super.init(frame: .zero)
    layer.addSublayer(previewLayer)
    backgroundColor = .black
  }
  required init?(coder: NSCoder) { fatalError("init(coder:) not supported") }
  override func layoutSubviews() {
    super.layoutSubviews()
    previewLayer.frame = bounds
  }
}

/// One native scanner instance hosted in a Flutter PlatformView. Wires the
/// per-view method/event channels and delegates camera work to `CameraManager`
/// and decoding to `FrameAnalyzer`.
final class ScannerPlatformView: NSObject, FlutterPlatformView {
  private let container: PreviewContainerView
  private let methodChannel: FlutterMethodChannel
  private let eventChannel: FlutterEventChannel
  private let events = EventDispatcher()

  private var config: ScannerConfig
  private var cameraManager: CameraManager?
  private var analyzer: FrameAnalyzer?

  init(frame: CGRect, viewId: Int64, args: Any?, messenger: FlutterBinaryMessenger) {
    let params = (args as? [String: Any?]) ?? [:]
    self.config = ScannerConfig.from(params)
    self.methodChannel = FlutterMethodChannel(
      name: Channels.methods(viewId), binaryMessenger: messenger)
    self.eventChannel = FlutterEventChannel(
      name: Channels.events(viewId), binaryMessenger: messenger)

    let analyzer = FrameAnalyzer(
      config: config,
      onBarcodes: { _ in },  // replaced below once events is captured
      onError: { _ in })
    let manager = CameraManager(analyzer: analyzer, events: events)
    self.analyzer = analyzer
    self.cameraManager = manager
    self.container = PreviewContainerView(previewLayer: manager.previewLayer)
    super.init()

    eventChannel.setStreamHandler(events)
    methodChannel.setMethodCallHandler { [weak self] call, result in
      self?.handle(call, result)
    }
  }

  func view() -> UIView { container }

  private func hasPermission() -> Bool {
    AVCaptureDevice.authorizationStatus(for: .video) == .authorized
  }

  private func handle(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
    switch call.method {
    case Method.initialize: initialize(call, result)
    case Method.start:
      cameraManager?.resumeAnalysis()
      events.send(["type": EventType.scannerStarted])
      result(nil)
    case Method.stop:
      cameraManager?.stop()
      events.send(["type": EventType.scannerStopped])
      result(nil)
    case Method.pause:
      cameraManager?.pauseAnalysis()
      result(nil)
    case Method.resume:
      cameraManager?.resumeAnalysis()
      result(nil)
    case Method.setFlash:
      let on = ((call.arguments as? [String: Any])?["enabled"] as? Bool) ?? false
      cameraManager?.setTorch(on)
      result(nil)
    case Method.toggleFlash:
      result(cameraManager?.toggleTorch() ?? false)
    case Method.switchCamera:
      cameraManager?.switchCamera(config: config)
      result(nil)
    case Method.setZoom:
      let z = ((call.arguments as? [String: Any])?["zoom"] as? NSNumber)?.doubleValue ?? 0
      cameraManager?.setZoom(z)
      result(nil)
    case Method.setExposure:
      let e = ((call.arguments as? [String: Any])?["exposure"] as? NSNumber)?.doubleValue ?? 0
      cameraManager?.setExposure(e)
      result(nil)
    case Method.setFocus:
      let a = call.arguments as? [String: Any]
      let x = (a?["x"] as? NSNumber)?.doubleValue ?? 0.5
      let y = (a?["y"] as? NSNumber)?.doubleValue ?? 0.5
      cameraManager?.focusAt(nx: x, ny: y)
      result(nil)
    case Method.captureFrame:
      if let data = cameraManager?.captureFrame() {
        result(FlutterStandardTypedData(bytes: data))
      } else {
        result(nil)
      }
    case Method.dispose:
      disposeInternal()
      result(nil)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  private func initialize(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
    if let params = call.arguments as? [String: Any?] {
      config = ScannerConfig.from(params)
    }
    guard hasPermission() else {
      events.send(["type": EventType.permissionDenied])
      result(FlutterError(
        code: ErrorCode.permissionDenied, message: "Camera permission not granted",
        details: nil))
      return
    }
    // Rebuild the analyzer/manager with a config-bound emission closure.
    let analyzer = FrameAnalyzer(
      config: config,
      onBarcodes: { [weak self] list in
        self?.events.send(["type": EventType.barcodeDetected, "barcodes": list])
      },
      onError: { [weak self] msg in
        self?.events.sendError(ErrorCode.decodingError, msg)
      })
    let manager = CameraManager(analyzer: analyzer, events: events)
    self.analyzer = analyzer
    self.cameraManager = manager
    container.layer.sublayers?.forEach { $0.removeFromSuperlayer() }
    container.layer.addSublayer(manager.previewLayer)
    manager.previewLayer.frame = container.bounds

    manager.start(
      config: config,
      onReady: { result(nil) },
      onError: { code, msg in
        self.events.sendError(code, msg)
        result(FlutterError(code: code, message: msg, details: nil))
      })
  }

  private func disposeInternal() {
    cameraManager?.dispose()
    cameraManager = nil
    analyzer = nil
    events.dispose()
    methodChannel.setMethodCallHandler(nil)
    eventChannel.setStreamHandler(nil)
  }

  deinit { disposeInternal() }
}

File: android/src/main/kotlin/com/umda/barcode_scanner_pro/ScannerPlatformView.kt
package com.umda.barcode_scanner_pro

import android.Manifest
import android.content.Context
import android.content.pm.PackageManager
import android.view.View
import androidx.camera.view.PreviewView
import androidx.core.content.ContextCompat
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.LifecycleRegistry
import com.google.mlkit.vision.barcode.BarcodeScannerOptions
import com.google.mlkit.vision.barcode.BarcodeScanning
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.platform.PlatformView

/**
 * A single native scanner instance hosted inside a Flutter PlatformView.
 *
 * Responsibilities are intentionally thin: it wires the per-view method/event
 * channels, owns a [PreviewView], and delegates all camera work to
 * [CameraManager] and all decoding to [FrameAnalyzer]. It also acts as a
 * self-driven [LifecycleOwner] so CameraX can bind its use cases.
 */
internal class ScannerPlatformView(
    private val context: Context,
    id: Int,
    creationParams: Map<String, Any?>?,
    messenger: BinaryMessenger,
) : PlatformView, MethodChannel.MethodCallHandler, LifecycleOwner {

    private val previewView = PreviewView(context).apply {
        implementationMode = PreviewView.ImplementationMode.COMPATIBLE
        scaleType = PreviewView.ScaleType.FILL_CENTER
    }

    private val methodChannel = MethodChannel(messenger, Channels.methods(id))
    private val eventChannel = EventChannel(messenger, Channels.events(id))
    private val events = EventDispatcher()
    private val registry = LifecycleRegistry(this)

    private var config: ScannerConfig =
        ScannerConfig.fromMap(creationParams ?: emptyMap())
    private var cameraManager: CameraManager? = null
    private var feedback: FeedbackController? = null

    init {
        methodChannel.setMethodCallHandler(this)
        eventChannel.setStreamHandler(events)
        registry.currentState = Lifecycle.State.CREATED
    }

    override fun getView(): View = previewView
    override val lifecycle: Lifecycle get() = registry

    private fun hasCameraPermission(): Boolean =
        ContextCompat.checkSelfPermission(context, Manifest.permission.CAMERA) ==
            PackageManager.PERMISSION_GRANTED

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        try {
            when (call.method) {
                Method.INITIALIZE -> initialize(call, result)
                Method.START -> {
                    cameraManager?.resumeAnalysis()
                    events.send(mapOf("type" to EventType.SCANNER_STARTED))
                    result.success(null)
                }
                Method.STOP -> {
                    cameraManager?.stop()
                    events.send(mapOf("type" to EventType.SCANNER_STOPPED))
                    result.success(null)
                }
                Method.PAUSE -> {
                    cameraManager?.pauseAnalysis()
                    result.success(null)
                }
                Method.RESUME -> {
                    cameraManager?.resumeAnalysis()
                    result.success(null)
                }
                Method.SET_FLASH -> {
                    cameraManager?.setTorch(call.argument<Boolean>("enabled") ?: false)
                    result.success(null)
                }
                Method.TOGGLE_FLASH -> result.success(cameraManager?.toggleTorch() ?: false)
                Method.SWITCH_CAMERA -> {
                    cameraManager?.switchCamera(config)
                    result.success(null)
                }
                Method.SET_ZOOM -> {
                    cameraManager?.setZoom((call.argument<Double>("zoom") ?: 0.0).toFloat())
                    result.success(null)
                }
                Method.SET_EXPOSURE -> {
                    cameraManager?.setExposure((call.argument<Double>("exposure") ?: 0.0).toFloat())
                    result.success(null)
                }
                Method.SET_FOCUS -> {
                    val x = (call.argument<Double>("x") ?: 0.5).toFloat()
                    val y = (call.argument<Double>("y") ?: 0.5).toFloat()
                    cameraManager?.focusAt(x, y)
                    result.success(null)
                }
                Method.CAPTURE_FRAME -> result.success(cameraManager?.captureFrame())
                Method.DISPOSE -> {
                    disposeInternal()
                    result.success(null)
                }
                else -> result.notImplemented()
            }
        } catch (e: Exception) {
            result.error(ErrorCode.DECODING_ERROR, e.localizedMessage, null)
        }
    }

    @Suppress("UNCHECKED_CAST")
    private fun initialize(call: MethodCall, result: MethodChannel.Result) {
        (call.arguments as? Map<String, Any?>)?.let { config = ScannerConfig.fromMap(it) }

        if (!hasCameraPermission()) {
            events.send(mapOf("type" to EventType.PERMISSION_DENIED))
            result.error(ErrorCode.PERMISSION_DENIED, "Camera permission not granted", null)
            return
        }

        val options = BarcodeScannerOptions.Builder()
            .setBarcodeFormats(FormatMapper.toMlKitFlags(config.formatMask))
            .build()
        val scanner = BarcodeScanning.getClient(options)

        feedback?.dispose()
        feedback = FeedbackController(
            context = context,
            soundEnabled = config.enableSound,
            vibrationEnabled = config.enableVibration,
        )

        val analyzer = FrameAnalyzer(
            scanner = scanner,
            config = config,
            onBarcodes = { list ->
                feedback?.onDetection()
                events.send(mapOf("type" to EventType.BARCODE_DETECTED, "barcodes" to list))
            },
            onError = { msg -> events.sendError(ErrorCode.DECODING_ERROR, msg) },
        )

        val manager = CameraManager(context, previewView, this, analyzer, events)
        cameraManager = manager
        registry.currentState = Lifecycle.State.RESUMED

        manager.start(
            config = config,
            onReady = { result.success(null) },
            onError = { code, msg ->
                events.sendError(code, msg)
                result.error(code, msg, null)
            },
        )
    }

    private fun disposeInternal() {
        cameraManager?.dispose()
        cameraManager = null
        feedback?.dispose()
        feedback = null
        registry.currentState = Lifecycle.State.DESTROYED
        events.dispose()
        methodChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
    }

    override fun dispose() = disposeInternal()
}
